tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/UninterpretedLiteralCache.scala

```diff
@@ -0,0 +1,72 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.types.{CellT, CellTFrom}
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, PureArena}
+import at.forsyte.apalache.tla.lir.{ConstT1, StrT1}
+import at.forsyte.apalache.tla.types.tla
+
+/**
+ * A cache for uninterpreted literals, that are translated as uninterpreted SMT constants, with a unique sort per
+ * uninterpreted type. Since two values are equal iff they are literally the same literal, we force inequality between
+ * all the respective SMT constants.
+ *
+ * Note that Strings are just a special kind of uninterpreted type.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class UninterpretedLiteralCache extends Cache[PureArena, (String, String), ArenaCell] {
+
+  /**
+   * Given a pair `(utype,idx)`, where `utype` represents an uninterpreted type name (possibly "Str") and `idx` some
+   * unique index within that type, returns an extension of `arena`, containing a cell, which represents
+   * "$idx_OF_$utype" (or "idx", if utype = "Str"), and said cell.
+   *
+   * Note that two values are equal (and get cached to the same cell) iff they have the same type and the same index, so
+   * e.g. "1_OF_A" and "1_OF_B" (passed here as ("A", "1") and ("B", "1")) get cached to different, incomparable cells,
+   * despite having the same index "1".
+   */
+  protected def create(
+      arena: PureArena,
+      typeAndIndex: (String, String)): (PureArena, ArenaCell) = {
+    // introduce a new cell
+    val (utype, _) = typeAndIndex
+    val cellType = if (utype == StrT1.toString) StrT1 else ConstT1(utype)
```

Comment:
I thought about this, and the problem is, that if it was `(TlaType1, Str)`, you'd have to manually filter out nonsense like `(IntT1, ...)`, since `StrT1` and `ConstT1` don't have a common superclass narrower than TT1. This way, the regex check on ConstT1 handles the type-string validation for us.

Reply:
When you say "you'd have to manually filter out nonsense like" could you help me understand what manually filtering means in this context? That is, is it just a matter of a sentinel check like

```
utype match {
  case StrT1 | ConstT1(_) => ..
  case _ => invalid type
}
```

or something more involved? It's just that I'm really surprised that at this point in the code we'd still be casting types to and from strings for comparison.

Reply:
I'll change it, since you seem to have a strong preference, I just didn't want to break the interface too much.